File: python/tests/test_cli_overlay.py
"""
Tests for YE.8: YAML Overlay Merge Support in CLI.

This module tests the --overlay / -f and --dump-merged CLI options
for the `tea run` command.
"""

import os
import re
import unittest
from pathlib import Path
from typing import Any

import yaml

from typer.testing import CliRunner

from the_edge_agent.cli import app
from the_edge_agent.schema.deep_merge import deep_merge, merge_all


# Path to test fixtures
FIXTURES_DIR = Path(__file__).parent.parent.parent / "tests" / "fixtures" / "overlay"

runner = CliRunner()


def strip_ansi(text: str) -> str:
    """Remove ANSI escape codes from text.

    Rich/Typer adds escape codes around option names which can break
    string matching in tests (e.g., '--dump-merged' becomes
    '\\x1b[1m-\\x1b[0m\\x1b[1m-dump\\x1b[0m\\x1b[1m-merged\\x1b[0m').
    """
    ansi_escape = re.compile(r"\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])")
    return ansi_escape.sub("", text)


class TestDeepMergeAlgorithm(unittest.TestCase):
    """Test the deep merge algorithm with kubectl-style semantics."""

    def test_object_recursive_merge(self):
        """Objects are recursively merged."""
        base = {"a": 1, "b": {"x": 10, "y": 20}}
        overlay = {"b": {"y": 30, "z": 40}, "c": 3}
        result = deep_merge(base, overlay)
        expected = {"a": 1, "b": {"x": 10, "y": 30, "z": 40}, "c": 3}
        self.assertEqual(result, expected)

    def test_array_replacement_not_concat(self):
        """Arrays are replaced, not concatenated."""
        base = {"items": [1, 2, 3]}
        overlay = {"items": [4, 5]}
        result = deep_merge(base, overlay)
        assert result is not None
        self.assertEqual(result["items"], [4, 5])

    def test_scalar_override(self):
        """Scalars use last-wins semantics."""
        base = {"count": 10, "name": "base"}
        overlay = {"count": 20}
        result = deep_merge(base, overlay)
        assert result is not None
        self.assertEqual(result["count"], 20)
        self.assertEqual(result["name"], "base")

    def test_null_can_override(self):
        """Null/None can override non-null values."""
        base = {"enabled": True, "value": 42}
        overlay = {"enabled": None}
        result = deep_merge(base, overlay)
        assert result is not None
        self.assertIsNone(result["enabled"])
        self.assertEqual(result["value"], 42)

    def test_nested_object_merge(self):
        """Deeply nested objects are merged correctly."""
        base = {"a": {"b": {"c": {"d": 1}}}}
        overlay = {"a": {"b": {"c": {"e": 2}}}}
        result = deep_merge(base, overlay)
        assert result is not None
        self.assertEqual(result["a"]["b"]["c"]["d"], 1)  # type: ignore[index]
        self.assertEqual(result["a"]["b"]["c"]["e"], 2)  # type: ignore[index]

    def test_type_mismatch_overlay_wins(self):
        """When types mismatch, overlay wins."""
        base = {"value": {"nested": True}}
        overlay = {"value": "string"}
        result = deep_merge(base, overlay)
        assert result is not None
        self.assertEqual(result["value"], "string")


class TestMergeAll(unittest.TestCase):
    """Test merging multiple schemas in order."""

    def test_multiple_overlays_in_order(self):
        """Later overlays override earlier ones."""
        schemas: list[dict[str, Any] | None] = [
            {"a": 1, "b": 2},
            {"b": 3, "c": 4},
            {"c": 5, "d": 6},
        ]
        result = merge_all(schemas)
        self.assertEqual(result, {"a": 1, "b": 3, "c": 5, "d": 6})

    def test_empty_list_returns_empty_dict(self):
        """Empty list returns empty dict."""
        result = merge_all([])
        self.assertEqual(result, {})

    def test_single_schema_returns_copy(self):
        """Single schema returns a copy."""
        schema = {"a": 1}
        result = merge_all([schema])
        self.assertEqual(result, {"a": 1})
        # Verify it's a copy, not the same object
        self.assertIsNot(result, schema)


class TestCLIOverlayOptions(unittest.TestCase):
    """Test CLI --overlay / -f options."""

    def test_help_shows_overlay_option(self):
        """--help shows the --overlay option."""
        result = runner.invoke(app, ["run", "--help"])
        output = strip_ansi(result.output)
        self.assertIn("--overlay", output)
        self.assertIn("-f", output)

    def test_help_shows_dump_merged_option(self):
        """--help shows the --dump-merged option."""
        result = runner.invoke(app, ["run", "--help"])
        output = strip_ansi(result.output)
        self.assertIn("--dump-merged", output)

    def test_overlay_flag_short_form(self):
        """Short form -f works."""
        base = FIXTURES_DIR / "base.yaml"
        overlay = FIXTURES_DIR / "overlay_settings.yaml"
        result = runner.invoke(
            app, ["run", str(base), "-f", str(overlay), "--dump-merged"]
        )
        self.assertEqual(result.exit_code, 0)
        self.assertIn("duckdb", result.output)  # From overlay

    def test_overlay_flag_long_form(self):
        """Long form --overlay works."""
        base = FIXTURES_DIR / "base.yaml"
        overlay = FIXTURES_DIR / "overlay_settings.yaml"
        result = runner.invoke(
            app, ["run", str(base), "--overlay", str(overlay), "--dump-merged"]
        )
        self.assertEqual(result.exit_code, 0)
        self.assertIn("duckdb", result.output)


class TestDumpMerged(unittest.TestCase):
    """Test --dump-merged output mode."""

    def test_dump_merged_outputs_yaml(self):
        """--dump-merged outputs valid YAML."""
        base = FIXTURES_DIR / "base.yaml"
        overlay = FIXTURES_DIR / "overlay_settings.yaml"
        result = runner.invoke(
            app, ["run", str(base), "-f", str(overlay), "--dump-merged"]
        )
        self.assertEqual(result.exit_code, 0)
        # Verify it's valid YAML
        parsed = yaml.safe_load(result.output)
        self.assertIsInstance(parsed, dict)
        self.assertEqual(parsed["name"], "test-agent")  # From base
        self.assertEqual(parsed["settings"]["model"], "gpt-4o")  # From overlay
        self.assertEqual(parsed["settings"]["ltm"]["backend"], "duckdb")  # From overlay
        self.assertEqual(
            parsed["settings"]["ltm"]["path"], "/tmp/tea_memory/"
        )  # From base

    def test_dump_merged_exits_without_running(self):
        """--dump-merged exits without executing the workflow."""
        base = FIXTURES_DIR / "base.yaml"
        result = runner.invoke(app, ["run", str(base), "--dump-merged"])
        self.assertEqual(result.exit_code, 0)
        # Should just output YAML, not run the workflow
        self.assertNotIn("Running agent", result.output)
        self.assertNotIn("Completed", result.output)

    def test_dump_merged_without_overlay(self):
        """--dump-merged without overlay outputs base config."""
        base = FIXTURES_DIR / "base.yaml"
        result = runner.invoke(app, ["run", str(base), "--dump-merged"])
        self.assertEqual(result.exit_code, 0)
        parsed = yaml.safe_load(result.output)
        self.assertEqual(parsed["settings"]["ltm"]["backend"], "sqlite")
        self.assertEqual(parsed["settings"]["model"], "gpt-4o-mini")


class TestMultipleOverlays(unittest.TestCase):
    """Test applying multiple overlays in order."""

    def test_multiple_overlays_applied_in_order(self):
        """Multiple overlays are applied in order (last wins)."""
        base = FIXTURES_DIR / "base.yaml"
        overlay1 = FIXTURES_DIR / "overlay_settings.yaml"
        overlay2 = FIXTURES_DIR / "overlay_variables.yaml"
        result = runner.invoke(
            app,
            [
                "run",
                str(base),
                "-f",
                str(overlay1),
                "-f",
                str(overlay2),
                "--dump-merged",
            ],
        )
        self.assertEqual(result.exit_code, 0)
        parsed = yaml.safe_load(result.output)
        # From overlay1
        self.assertEqual(parsed["settings"]["model"], "gpt-4o")
        # From overlay2
        self.assertEqual(parsed["variables"]["greeting"], "bonjour")
        self.assertEqual(parsed["variables"]["new_var"], "added_by_overlay")

    def test_later_overlay_overrides_earlier(self):
        """Later overlay merges nodes by key (YE.9 behavior)."""
        base = FIXTURES_DIR / "base.yaml"
        overlay_nodes = FIXTURES_DIR / "overlay_nodes.yaml"
        result = runner.invoke(
            app, ["run", str(base), "-f", str(overlay_nodes), "--dump-merged"]
        )
        self.assertEqual(result.exit_code, 0)
        parsed = yaml.safe_load(result.output)
        # YE.9: Nodes merge by name key - base has "process", overlay has "custom_process"
        # Since names don't match, custom_process is appended
        self.assertEqual(len(parsed["nodes"]), 2)
        node_names = [n["name"] for n in parsed["nodes"]]
        self.assertIn("process", node_names)  # Preserved from base
        self.assertIn("custom_process", node_names)  # Added from overlay


class TestErrorHandling(unittest.TestCase):
    """Test error handling for overlay operations."""

    def test_missing_overlay_file_error(self):
        """Clear error for missing overlay file."""
        base = FIXTURES_DIR / "base.yaml"
        result = runner.invoke(
            app, ["run", str(base), "-f", "/nonexistent/overlay.yaml", "--dump-merged"]
        )
        self.assertNotEqual(result.exit_code, 0)
        self.assertIn("Overlay file not found", result.output)
        self.assertIn("/nonexistent/overlay.yaml", result.output)

    def test_invalid_yaml_parse_error(self):
        """Clear error for invalid YAML in overlay."""
        base = FIXTURES_DIR / "base.yaml"
        invalid = FIXTURES_DIR / "invalid.yaml"
        result = runner.invoke(
            app, ["run", str(base), "-f", str(invalid), "--dump-merged"]
        )
        self.assertNotEqual(result.exit_code, 0)
        self.assertIn("Invalid YAML", result.output)

    def test_missing_base_file_error(self):
        """Clear error for missing base file."""
        overlay = FIXTURES_DIR / "overlay_settings.yaml"
        result = runner.invoke(
            app, ["run", "/nonexistent/base.yaml", "-f", str(overlay), "--dump-merged"]
        )
        self.assertNotEqual(result.exit_code, 0)
        self.assertIn("not found", result.output)


class TestEmptyOverlay(unittest.TestCase):
    """Test handling of empty overlay files."""

    def test_empty_overlay_file(self):
        """Empty overlay file is treated as empty dict."""
        base = FIXTURES_DIR / "base.yaml"
        empty = FIXTURES_DIR / "empty.yaml"
        result = runner.invoke(
            app, ["run", str(base), "-f", str(empty), "--dump-merged"]
        )
        self.assertEqual(result.exit_code, 0)
        parsed = yaml.safe_load(result.output)
        # Should match base config exactly
        self.assertEqual(parsed["settings"]["ltm"]["backend"], "sqlite")


class TestNullOverride(unittest.TestCase):
    """Test null/None override behavior."""

    def test_null_can_override_value(self):
        """Null in overlay can override non-null base value."""
        base = FIXTURES_DIR / "base.yaml"
        null_overlay = FIXTURES_DIR / "overlay_null.yaml"
        result = runner.invoke(
            app, ["run", str(base), "-f", str(null_overlay), "--dump-merged"]
        )
        self.assertEqual(result.exit_code, 0)
        parsed = yaml.safe_load(result.output)
        # temperature should be null/None
        self.assertIsNone(parsed["settings"]["temperature"])


class TestOverlayWithOtherFlags(unittest.TestCase):
    """Test that overlay works with other CLI flags."""

    def test_overlay_with_input_flag(self):
        """--overlay works with --input flag."""
        base = FIXTURES_DIR / "base.yaml"
        overlay = FIXTURES_DIR / "overlay_settings.yaml"
        # This should parse successfully (we use --dump-merged to avoid execution)
        result = runner.invoke(
            app,
            [
                "run",
                str(base),
                "-f",
                str(overlay),
                "--input",
                '{"test": "value"}',
                "--dump-merged",
            ],
        )
        self.assertEqual(result.exit_code, 0)

    def test_overlay_preserves_base_name(self):
        """Overlay preserves name from base unless overridden."""
        base = FIXTURES_DIR / "base.yaml"
        overlay = FIXTURES_DIR / "overlay_settings.yaml"
        result = runner.invoke(
            app, ["run", str(base), "-f", str(overlay), "--dump-merged"]
        )
        self.assertEqual(result.exit_code, 0)
        parsed = yaml.safe_load(result.output)
        self.assertEqual(parsed["name"], "test-agent")  # From base


class TestImportsMerge(unittest.TestCase):
    """Test imports section merging."""

    def test_imports_preserved_from_base(self):
        """Imports from base are preserved."""
        # Create a temporary base with imports
        import tempfile

        with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml", delete=False) as f:
            yaml.dump(
                {
                    "name": "test",
                    "imports": [{"path": "base_action.py", "namespace": "base"}],
                    "nodes": [{"name": "test", "run": "return {}"}],
                    "edges": [
                        {"from": "__start__", "to": "test"},
                        {"from": "test", "to": "__end__"},
                    ],
                },
                f,
            )
            base_path = f.name

        with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml", delete=False) as f:
            yaml.dump(
                {"imports": [{"path": "overlay_action.py", "namespace": "overlay"}]}, f
            )
            overlay_path = f.name

        try:
            result = runner.invoke(
                app, ["run", base_path, "-f", overlay_path, "--dump-merged"]
            )
            self.assertEqual(result.exit_code, 0)
            parsed = yaml.safe_load(result.output)
            # Arrays are replaced, so only overlay imports should exist
            self.assertEqual(len(parsed["imports"]), 1)
            self.assertEqual(parsed["imports"][0]["namespace"], "overlay")
        finally:
            os.unlink(base_path)
            os.unlink(overlay_path)


if __name__ == "__main__":
    unittest.main()

File: rust/tests/test_cli_overlay.rs
//! CLI Overlay Tests - YE.8: YAML Overlay Merge Support
//!
//! These tests verify the --overlay / -f and --dump-merged CLI options
//! for the `tea run` command.

use assert_cmd::Command;
use predicates::prelude::*;
use std::path::PathBuf;

/// Get path to test fixtures
fn fixtures_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .join("tests/fixtures/overlay")
}

/// Get the tea binary
fn tea_cmd() -> Command {
    #[allow(deprecated)]
    Command::cargo_bin("tea").unwrap()
}

// ============================================================================
// CLI Options Tests
// ============================================================================

#[test]
fn test_help_shows_overlay_option() {
    tea_cmd()
        .args(["run", "--help"])
        .assert()
        .success()
        .stdout(predicate::str::contains("--overlay"))
        .stdout(predicate::str::contains("-f"));
}

#[test]
fn test_help_shows_dump_merged_option() {
    tea_cmd()
        .args(["run", "--help"])
        .assert()
        .success()
        .stdout(predicate::str::contains("--dump-merged"));
}

// ============================================================================
// Dump Merged Tests
// ============================================================================

#[test]
fn test_dump_merged_outputs_yaml() {
    let base = fixtures_dir().join("base.yaml");
    let overlay = fixtures_dir().join("overlay_settings.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            overlay.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("name: test-agent")) // From base
        .stdout(predicate::str::contains("model: gpt-4o")) // From overlay
        .stdout(predicate::str::contains("duckdb")); // From overlay
}

#[test]
fn test_dump_merged_without_overlay() {
    let base = fixtures_dir().join("base.yaml");

    tea_cmd()
        .args(["run", base.to_str().unwrap(), "--dump-merged"])
        .assert()
        .success()
        .stdout(predicate::str::contains("sqlite")) // From base
        .stdout(predicate::str::contains("gpt-4o-mini")); // From base
}

#[test]
fn test_dump_merged_exits_without_running() {
    let base = fixtures_dir().join("base.yaml");

    // Should just output YAML, not run the workflow
    tea_cmd()
        .args(["run", base.to_str().unwrap(), "--dump-merged"])
        .assert()
        .success()
        .stdout(predicate::str::contains("name:"))
        .stdout(
            predicate::str::is_match(r"Running|Completed|executing")
                .unwrap()
                .not(),
        );
}

// ============================================================================
// Multiple Overlays Tests
// ============================================================================

#[test]
fn test_multiple_overlays_applied_in_order() {
    let base = fixtures_dir().join("base.yaml");
    let overlay1 = fixtures_dir().join("overlay_settings.yaml");
    let overlay2 = fixtures_dir().join("overlay_variables.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            overlay1.to_str().unwrap(),
            "-f",
            overlay2.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("gpt-4o")) // From overlay1
        .stdout(predicate::str::contains("bonjour")) // From overlay2
        .stdout(predicate::str::contains("added_by_overlay")); // From overlay2
}

#[test]
fn test_later_overlay_overrides_earlier() {
    let base = fixtures_dir().join("base.yaml");
    let overlay_nodes = fixtures_dir().join("overlay_nodes.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            overlay_nodes.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success()
        // Nodes array should be completely replaced (arrays replace, not merge)
        .stdout(predicate::str::contains("custom_process"))
        .stdout(predicate::str::contains("CUSTOM:"));
}

// ============================================================================
// Error Handling Tests
// ============================================================================

#[test]
fn test_missing_overlay_file_error() {
    let base = fixtures_dir().join("base.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            "/nonexistent/overlay.yaml",
            "--dump-merged",
        ])
        .assert()
        .failure()
        .stderr(predicate::str::contains("Overlay file not found"));
}

#[test]
fn test_invalid_yaml_parse_error() {
    let base = fixtures_dir().join("base.yaml");
    let invalid = fixtures_dir().join("invalid.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            invalid.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .failure()
        .stderr(predicate::str::contains("Invalid YAML"));
}

#[test]
fn test_missing_base_file_error() {
    let overlay = fixtures_dir().join("overlay_settings.yaml");

    tea_cmd()
        .args([
            "run",
            "/nonexistent/base.yaml",
            "-f",
            overlay.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .failure()
        .stderr(predicate::str::contains("Failed to read"));
}

// ============================================================================
// Empty Overlay Tests
// ============================================================================

#[test]
fn test_empty_overlay_file() {
    let base = fixtures_dir().join("base.yaml");
    let empty = fixtures_dir().join("empty.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            empty.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success()
        // Should match base config exactly
        .stdout(predicate::str::contains("sqlite"))
        .stdout(predicate::str::contains("gpt-4o-mini"));
}

// ============================================================================
// Null Override Tests
// ============================================================================

#[test]
fn test_null_can_override_value() {
    let base = fixtures_dir().join("base.yaml");
    let null_overlay = fixtures_dir().join("overlay_null.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            null_overlay.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success()
        // temperature should be null
        .stdout(
            predicate::str::contains("temperature: null")
                .or(predicate::str::contains("temperature: ~")),
        );
}

// ============================================================================
// Flag Combination Tests
// ============================================================================

#[test]
fn test_overlay_short_flag_works() {
    let base = fixtures_dir().join("base.yaml");
    let overlay = fixtures_dir().join("overlay_settings.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            overlay.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success();
}

#[test]
fn test_overlay_long_flag_works() {
    let base = fixtures_dir().join("base.yaml");
    let overlay = fixtures_dir().join("overlay_settings.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "--overlay",
            overlay.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success();
}

#[test]
fn test_overlay_preserves_base_name() {
    let base = fixtures_dir().join("base.yaml");
    let overlay = fixtures_dir().join("overlay_settings.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            overlay.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("name: test-agent")); // From base
}

// ============================================================================
// Deep Merge Verification Tests
// ============================================================================

#[test]
fn test_nested_settings_merge() {
    let base = fixtures_dir().join("base.yaml");
    let overlay = fixtures_dir().join("overlay_settings.yaml");

    tea_cmd()
        .args([
            "run",
            base.to_str().unwrap(),
            "-f",
            overlay.to_str().unwrap(),
            "--dump-merged",
        ])
        .assert()
        .success()
        // Nested ltm settings should be merged:
        // base.settings.ltm.path should still exist
        .stdout(predicate::str::contains("/tmp/tea_memory/"))
        // overlay.settings.ltm.backend should override
        .stdout(predicate::str::contains("duckdb"))
        // overlay.settings.ltm.catalog should be added
        .stdout(predicate::str::contains("firestore"));
}
